browser/ui/webui/brave_education/brave_education_ui_browsertest.cc

```diff
@@ -0,0 +1,192 @@
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include <string>
+#include <string_view>
+#include <utility>
+
+#include "base/memory/weak_ptr.h"
+#include "brave/components/brave_education/common/education_content_urls.h"
+#include "brave/components/brave_vpn/common/buildflags/buildflags.h"
+#include "brave/components/constants/webui_url_constants.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/chrome_pages.h"
+#include "chrome/common/webui_url_constants.h"
+#include "chrome/test/base/chrome_test_utils.h"
+#include "chrome/test/base/in_process_browser_test.h"
+#include "chrome/test/base/ui_test_utils.h"
+#include "content/public/browser/web_ui_data_source.h"
+#include "content/public/test/browser_test.h"
+#include "content/public/test/browser_test_utils.h"
+
+namespace brave_education {
+
+class BraveEducationUIBrowserTest : public InProcessBrowserTest {
+ protected:
+  void SetUpOnMainThread() override { CreateAndAddWebUITestDataSource(); }
+
+  void NavigateToEducationPage() {
+    auto content_type = EducationContentType::kGettingStartedAdBlock;
+    auto webui_url = GetEducationContentBrowserURL(content_type);
+    ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), webui_url));
+    auto* web_contents = chrome_test_utils::GetActiveWebContents(this);
+    ASSERT_TRUE(WaitForLoadStop(web_contents));
+  }
+
+  void PostMessageFromIFrame(std::string_view message_data) {
+    message_data_ = message_data;
+    auto* web_contents = chrome_test_utils::GetActiveWebContents(this);
+    ASSERT_TRUE(content::ExecJs(web_contents, R"(
+      const iframe = document.getElementById('content')
+      iframe.src = "chrome://webui-test/"
+    )"));
+  }
+
+ private:
+  void CreateAndAddWebUITestDataSource() {
+    auto* web_contents = chrome_test_utils::GetActiveWebContents(this);
+
+    content::WebUIDataSource* source = content::WebUIDataSource::CreateAndAdd(
+        web_contents->GetBrowserContext(), chrome::kChromeUIWebUITestHost);
+
+    source->OverrideContentSecurityPolicy(
+        network::mojom::CSPDirectiveName::FrameAncestors,
+        "frame-ancestors chrome://* 'self';");
+
+    source->SetRequestFilter(
+        base::BindRepeating([](const std::string&) { return true; }),
+        base::BindRepeating(
+            &BraveEducationUIBrowserTest::HandleWebUITestRequest,
+            weak_factory_.GetWeakPtr()));
+  }
+
+  void HandleWebUITestRequest(
+      const std::string& path,
+      content::WebUIDataSource::GotDataCallback callback) {
+    if (path == "post-message.js") {
+      std::string js = "window.parent.postMessage(" + message_data_ + ", '*')";
+      std::move(callback).Run(
+          base::MakeRefCounted<base::RefCountedString>(std::move(js)));
+      return;
+    }
+
+    static const char kHTML[] = R"(
```

Comment:
can we add a test here that if the page navigates to a different origin via something like `window.location.href`, commands do not run successfully?

Reply:
Based on the CSP restrictions, it will refuse to frame content from any origin besides `https://brave.com` or `chrome://webui-test` (even through clicking a link) so a test can't be written in this way.